packages/grafana-data/src/transformations/transformers/reduce.ts

```diff
@@ -33,14 +37,55 @@ export const reduceTransformer: DataTransformerInfo<ReduceTransformerOptions> =
   operator: options => source =>
     source.pipe(
       map(data => {
-        const matcher = options.fields ? getFieldMatcher(options.fields) : alwaysFieldMatcher;
+        const matcher = options.fields
+          ? getFieldMatcher(options.fields)
+          : options.includeTimeField && options.mode === ReduceTransformerMode.ReduceFields
+          ? alwaysFieldMatcher
+          : notTimeFieldMatcher;
         const calculators = options.reducers && options.reducers.length ? fieldReducers.list(options.reducers) : [];
         const reducers = calculators.map(c => c.id);
 
         const processed: DataFrame[] = [];
+        // Collapse all matching fields into a single row
+        if (options.mode === ReduceTransformerMode.ReduceFields) {
+          for (const series of data) {
+            const fields: Field[] = [];
+            for (const field of series.fields) {
+              if (matcher(field, series, data)) {
+                const results = reduceField({
+                  field,
+                  reducers,
+                });
+                for (const reducer of reducers) {
+                  const value = results[reducer];
+                  const copy = {
+                    ...field,
+                    values: new ArrayVector([value]),
+                  };
+                  copy.state = undefined;
+                  if (reducers.length > 1) {
+                    if (!copy.labels) {
+                      copy.labels = {};
+                    }
+                    copy.labels['reducer'] = fieldReducers.get(reducer).name;
+                  }
+                  fields.push(copy);
+                }
+              }
+            }
+            if (fields.length) {
+              processed.push({
+                ...series,
+                fields,
+                length: 1, // always one row
+              });
+            }
+          }
+
+          return processed;
+        }
```

Comment:
This reducer function is 120 lines of pretty dense code. Would be nice to be if we could break it up a bit. And nitpick: I struggle reading large code blocks like this that has no breaks between control statement, it all mixes and together and I mistake statements to look like they form a single expression but are really completely separate expressions.